Design of data preprocessing in AutoML.md

```diff
@@ -0,0 +1,93 @@
+### Goal
+Provide a way to support feature preprocessing within AutoML pipeline.
```

Comment:
The goal is to integrate Target Encoding, not necessarily a generic feature preprocessing. Generic feature preprocessing is one way to do it but is not a necessity.

Reply:
@michalkurka I think that here is an example of our main issue. We should not just follow these 3 steps:

1. Design
2. Review design
3. Implement

 as you mentioned in h2o-3-auto-ml channel  https://h2oai.slack.com/archives/C0E0ADTM1/p1580849237085900?thread_ts=1580839939.083900&cid=C0E0ADTM1

Steps instead should be:
1.  Discuss the goal with all interested parties (automl team, h2o core team, etc)
2. Design
3. Review design
4. Implement

Goal is seriously narrowing down space of possible solutions. In most cases it defines solution in 1:1 fashion.

In current situation we have two different goals for a single task;
1) AutoML team:  Integrate TE into AutoML and make it **in generic way**  ( here is a discussion with @sebhrusen https://github.com/h2oai/h2o-3/pull/4158#issuecomment-579401097 )
2) Core team:   Integrate TE into AutoML **in not necessarily generic way** ( here is a statement https://github.com/h2oai/h2o-3/pull/4303#discussion_r376519702)

So no one can "sit on multiple chairs". But I consider myself as a member of AutoML team so I should probably be in favour our AutoML team's goals. It is not clear to me whether we should always agree with core teams goals. But anyway let's make a |discussion of the goal" a first step. 
This is the main issue and not implementing without reviewed design(but of course this is a good step as well)
@ledell @sebhrusen  @mmalohlava

Reply:
@michalkurka @deil87 I would suggest that put both goals here.
1. One is narrow and more short-term: TE integration.
2. A broader one to fulfil AutoML extension needs. As Andrey mentioned, AutoML plans to add more feature engineering in a near future: feature extraction, selection and probably more...
If AutoML simply wanted to apply PCA today, it would not be possible to provide a usable model to the user: so on our side, the absence of such pipeline abstraction is impeding us.

So, if I agree that the current integration deals only with TE, we should not forget the bigger picture on AutoML side, and for such level of abstraction, I think it's reasonable to add it to the goal.
The fact that TE is currently both an encoding step and a model (like PCA) probably adds to the confusion between those 2 goals...

Reply:
@sebhrusen I agree that the bigger picture should always be considered.

However, the solution should IMHO be designed to perfectly solve the specific problem. It might happen that a generic solution is _the_ perfect solution we are looking for.

I am not concerned much with how things are implemented in the code (that can always change) but more of what is presented to the users. What does the solution take away and what are the advantages.

Reply:
@sebhrusen 
> The fact that TE is currently both an encoder and a model (like PCA) probably adds to the confusion between those 2 goals...

Could you please elaborate on what do you mean it is both? It is just Model like PCA. The fact that there is a helper TargetEncoder should not contribute to confusions. Goals imho are different just because what is long-term from h2o-core perspective is short-term from automl team's perspective. 

I also thought that storing references to other models in a cornerstone Model class should be considered as  not an ideal solution.

Reply:
@michalkurka @sebhrusen To formalise what was discussed here I have created jira for our long-term goal on Pipelines https://0xdata.atlassian.net/browse/PUBDEV-7291 .